ASP-based solver: tweak heuristic, modify compiler encoding

alalazo opened the pull request:
This PR does 2 unrelated things:
1. It changes the encoding of the compilers
2. It tweaks the heuristic for the solves in a0d88179074f81d13a3fad629a43d86334e7f982

Both were initially motivated by trying to get a performance gain but, while 2 showed significant speed-ups[^1], 1 instead didn't. I kept it anyhow, since I think the code related to compilers is more consolidated with the new encoding and we might get some performance improvement out of it if we can base our errors on the `node_compiler(Package, CompilerID)` atoms instead of `attrs`.

[^1]: In general the changes in the heuristic brought a ~10% speed-up on the tests I did. I'll post detailed results below.

alalazo committed:
Emit facts for target compatibility directly

This is just a simplification, and it doesn't seem to
affect performance in any way.

alalazo committed:
compilers: slightly simplify setup code

alalazo committed:
compilers: unify functions for compiler facts

alalazo committed:
compilers: rework compiler encoding (add a compiler ID)

alalazo committed:
Tweak heuristic and break symmetry for some performance gain

alalazo commented:
# Summary

- `--fresh` seems to be 10% faster in this branch, 
- `--reuse` has outliers which are 15% slower but is generally much faster (e.g. `zfp` is 7x faster)

# Details

All the tests below concretize the packages tagged as radiuss:

[radiuss.txt](https://github.com/spack/spack/files/10941368/radiuss.txt)

one by one under different conditions. This branch has been compared with Spack `develop` at 617f44f9eda73183eb29c6faae088790796ad8c8

## Concretize without reuse

For the first test the concretization was using `--fresh` and the timing obtained for both branches is:

[radiuss_develop.csv](https://github.com/spack/spack/files/10941418/radiuss_develop.csv)
[radiuss_heuristic.csv](https://github.com/spack/spack/files/10941417/radiuss_heuristic.csv)

**In general there is a slight improvement on this branch of around ~10%**.

The plots are on `develop`:

![develop](https://user-images.githubusercontent.com/4199709/224302534-ff6c7b87-6a4d-4ef3-9ad0-e5804aacbfd6.png)

and on this branch:

![heuristic](https://user-images.githubusercontent.com/4199709/224302584-c0da97dc-d25e-44a4-9a7c-7e64ad7581eb.png)


## Concretize with reuse

The second test has been done with `--reuse` after adding:
```console
spack mirror add e4s https://cache.e4s.io/23.02
```

Results are:

[radiuss_heuristic_mirror.csv](https://github.com/spack/spack/files/10941488/radiuss_heuristic_mirror.csv)
[radiuss_develop_mirror.csv](https://github.com/spack/spack/files/10941489/radiuss_develop_mirror.csv)



**There is an outlier that is ~15% slower on this branch, but others like `zfp` are 7x faster.**

The results on `develop` are:

![develop_mirror](https://user-images.githubusercontent.com/4199709/224303212-abf6e8f1-a456-4ea8-897e-fa19f351df0b.png)

and on this branch:

![heuristic_mirror](https://user-images.githubusercontent.com/4199709/224303254-e8c2ba58-6c23-41ed-a9f5-ff909dcbdf07.png)

alalazo commented:
@adamjstewart Can you try to concretize some of your environments and check if the results above are in line with what you see?

adamjstewart commented:
On my ancient laptop with 614 installed packages:

### Develop

```console
$ time spack spec --fresh py-torch
...
real	0m45.806s
user	0m43.379s
sys	0m1.748s

$ time spack spec --reuse py-torch
...
real	0m51.654s
user	0m48.387s
sys	0m1.930s
```

### This branch

```console
$ time spack spec --fresh py-torch
...
real	0m46.610s
user	0m43.943s
sys	0m1.814s

$ time spack spec --reuse py-torch
...
real	0m45.537s
user	0m43.283s
sys	0m1.643s
```

Trying to avoid reconcretizing my environments because chaos, but I can duplicate the environment if you want me to measure that too.

alalazo commented:
If you can post the `spack.yaml` I'll try to double check that.

adamjstewart commented:
[spack.txt](https://github.com/spack/spack/files/10943503/spack.txt)

It's a big boy.

alalazo commented:
Concretizing your environment on my laptop on `develop`:
```console
$ time spack -e . concretize --fresh
[ ... ]
real	1m46,189s
user	1m43,392s
sys	0m2,324s
```
and with this PR:
```console
$ time spack -e . concretize --fresh
[ ... ]
real	1m33,070s
user	1m30,739s
sys	0m2,098s
```

alalazo commented:
For reference on the encoding change (that in the end didn't affect performance). This is how a compiler looks like in this PR:
```
%----------------------------------------------------------------------------
% Available compilers
%----------------------------------------------------------------------------
compiler_id(0).
compiler_name(0,"oneapi").
compiler_version(0,"2023.0.0").
compiler_os(0,"ubuntu20.04").
compiler_target(0,"x86_64").
```
It has an integer ID and facts with arity 2 that attach a single property to that ID. Before we were repeating compiler name and compiler version in multiple places and had to match on them.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -942,54 +945,62 @@ def conflict_rules(self, pkg):
-            compiler_versions[compiler.name].add(compiler.version)
+            # FIXME: same spec with multiple OS and targets?
+            if key in self.compiler_info:
+                warnings.warn(
+                    f"compiler {key} has duplicate entries, only the first one will be considered"
+                )
```
I think this should just take the first one without warning.  It's common to have the same compiler spec with different OS/target settings.  See https://github.com/spack/spack/pull/35721

@becker33: do you agree?

alalazo replied:
> I think this should just take the first one without warning. 

I added the warning because right now taking the first one doesn't mean we take the _right_ one (the first entry for compiler spec `X@Y` might be using the wrong OS, or the wrong target). For instance with a configuration like:
```yaml
compilers:
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    operating_system: ubuntu20.04
    target: x86_64
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    operating_system: ubuntu20.04
    target: aarch64
```
the first entry is wrong if we are targeting an `aarch64`. 

---

Regarding flags, the new encoding is a step forward towards differentiating compilers, since the encoding for this:
```yaml
compilers:
- compiler:
    spec: gcc@9.4.0
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    flags:
      cflags: "-O2"
    operating_system: ubuntu20.04
    target: x86_64
```
is the following:
```
compiler_id(2).
compiler_name(2,"gcc").
compiler_version(2,"9.4.0").
compiler_os(2,"ubuntu20.04").
compiler_target(2,"x86_64").
compiler_flag(2,"cflags","-O2").
```
This means that the new encoding in ASP can differentiate two compiler with the same name and version but different flags. The warning is there since we have no way at the moment to select them from the spec syntax[^1].

[^1]: The workaround for this use case, as far as I know, is to tweak the version when another unselectable parameter is different.

alalazo replied:
The warning is now triggered if we have compilers which can't be distinguished using the spec syntax (same spec, os and target).

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -942,54 +945,62 @@ def conflict_rules(self, pkg):
+            self.gen.fact(fn.compiler_version(compiler_id, compiler.version))
+
+            compiler_entry = self.compiler_info[compiler]
+            # Use "get" to allow concretizing even when copiler existence checks are disabled
```
```suggestion
            # Use "get" to allow concretizing even when compiler existence checks are disabled
```

alalazo replied:
Fixed

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1931,14 +1919,12 @@ def versions_for(v):
                     self.gen.fact(
                         fn.compiler_version_satisfies(
-                            constraint.name, constraint.versions, compiler.version
+                            constraint.name, constraint.versions, compiler_id
```
This feels weird to me but I think it's the right thing.  I will be happier once compilers are proper packages.

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -868,32 +858,44 @@ error(2, "'{0} target={1}' is not compatible with this machine", Package, Target
+% Expand the internal attribute into "attr("node_compiler_version")
+attr("node_compiler_version", Package, CompilerName, CompilerVersion)
+  :- node_compiler(Package, CompilerID),
+     compiler_name(CompilerID, CompilerName),
+     compiler_version(CompilerID, CompilerVersion),
```
Would it make sense to remap `attr("node_compiler_version", Package, CompilerName, CompilerVersion)` to `attr("node_compiler_version", Package, CompilerID)`?  I can't decide if it helps... it requires a lookup in the `SpecBuilder` so I *think* I like it better the way it is here.

alalazo replied:
I can have a look at this in a following PR. It might help to do like you say and have a similar lookup on `error(2, ...)` involving compilers. But I can't tell for sure without trying it first (as anything clingo :sweat: )

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1311,13 +1327,29 @@ opt_criterion(5, "non-preferred targets").
-#heuristic version_weight(Package, 0) : version_declared(Package, Version, 0), attr("node", Package). [10, true]
-#heuristic attr("node_target", Package, Target) : package_target_weight(Target, Package, 0), attr("node", Package). [10, true]
-#heuristic node_target_weight(Package, 0) : attr("node", Package). [10, true]
+#heuristic literal_solved(ID) : literal(ID). [1, sign]
+#heuristic literal_solved(ID) : literal(ID). [50, init]
+#heuristic attr("hash", Package, Hash) : attr("root", Package). [45, init]
+
+#heuristic attr("version", Package, Version) : version_declared(Package, Version, 0), attr("root", Package). [40, true]
+#heuristic version_weight(Package, 0) : version_declared(Package, Version, 0), attr("root", Package). [40, true]
+#heuristic attr("variant_value", Package, Variant, Value) : variant_default_value(Package, Variant, Value), attr("root", Package). [40, true]
+#heuristic attr("node_target", Package, Target) : package_target_weight(Target, Package, 0), attr("root", Package). [40, true]
+#heuristic node_target_weight(Package, 0) : attr("root", Package). [40, true]
+#heuristic node_compiler(Package, CompilerID) : default_compiler_preference(ID, 0), compiler_id(ID), attr("root", Package). [40, true]
+
+#heuristic provider(Package, Virtual) : possible_provider_weight(Package, Virtual, 0, _), attr("virtual_node", Virtual). [30, true]
+#heuristic provider_weight(Package, Virtual, 0, R) : possible_provider_weight(Package, Virtual, 0, R), attr("virtual_node", Virtual). [30, true]
+#heuristic attr("node", Package) : possible_provider_weight(Package, Virtual, 0, _), attr("virtual_node", Virtual). [30, true]
+
+#heuristic attr("version", Package, Version) : version_declared(Package, Version, 0), attr("node", Package). [20, true]
+#heuristic version_weight(Package, 0) : version_declared(Package, Version, 0), attr("node", Package). [20, true]
+
+#heuristic attr("node_target", Package, Target) : package_target_weight(Target, Package, 0), attr("node", Package). [20, true]
+#heuristic node_target_weight(Package, 0) : attr("node", Package). [20, true]
+#heuristic node_compiler(Package, CompilerID) : default_compiler_preference(ID, 0), compiler_id(ID), attr("node", Package). [15, true]
+
```
Did you test how many of these actually help performance?  Do all of them?  It's a lot of heuristics, so I wonder if some can be removed. When I tried tweaking these heuristics for solves, I found that not all of them mattered, but that was for a limited set of tests.

alalazo replied:
Most of them are from before, they just shifted priority. The new ones are:
```
#heuristic literal_solved(ID) : literal(ID). [1, sign]
#heuristic literal_solved(ID) : literal(ID). [50, init]
#heuristic attr("hash", Package, Hash) : attr("root", Package). [45, init]
```
which basically mean to start the solve considering that everything is solvable, and trying hard at the beginning to reuse hashes (`init` is an initial "priority" multiplier that is damped as the solve algorithm proceeds). Then I also differentiated the `root` attributes from the rest of the DAG, and gave them higher priority.

alalazo replied:
Also, not sure this is a good approach or not but what I did was to try having heuristic for all the choices we have in the logic program. The rational is that everything else should be derivable from rules, once we have guesses for choices.

alalazo committed:
Fix typo in comment

alalazo committed:
Fix warning on compilers, simplify setup

The warning is now triggered only if there are
multiple compilers with the same spec, os and
target (since they can't be selected by users
with the spec syntax only).
